GC life time is shorter than transaction duration while running DDL

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. Run add index job in a large table, and set the batch_size to 10240.

It is not easy to reproduce because the DDL txn is usually fast.

<!-- a step by step guide for reproducing the bug. -->

### 2. What did you expect to see? (Required)

No error.

### 3. What did you see instead (Required)

Ingest add index job fallback to txn-merge mode. 
```
[INFO] [index.go:762] ["[ddl] fallback to txn-merge backfill process"] [error="[tikv:9006]GC life time is shorter than transaction duration, transaction starts at 2022-12-20 10:58:20.94 +0000 UTC, GC safe point is 2022-12-20 10:58:22.14 +0000 UTC"]
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

ed33fdea6

The problem still exist in b51758efdbd97:

```
[2023/05/08 07:34:36.517 +00:00] [ERROR] [backfilling_scheduler.go:461] ["[ddl-ingest] encounter error when handle index chunk"] [id=10736] [error="[tikv:9006]GC life time is shorter than transaction duration, transaction starts at 2023-05-08 07:17:32.674 +0000 UTC, GC safe point is 2023-05-08 07:19:07.624 +0000 UTC"]
```

--- a/ddl/BUILD.bazel
+++ b/ddl/BUILD.bazel
@@ -223,6 +223,7 @@ go_test(
         "//config",
         "//ddl/ingest",
         "//ddl/internal/callback",
+        "//ddl/internal/session",
         "//ddl/placement",
         "//ddl/schematracker",
         "//ddl/testutil",

--- a/ddl/backfilling_scheduler.go
+++ b/ddl/backfilling_scheduler.go
@@ -434,7 +434,7 @@ func (b *ingestBackfillScheduler) createCopReqSenderPool() (*copReqSenderPool, e
 		logutil.BgLogger().Warn("[ddl-ingest] cannot init cop request sender", zap.Error(err))
 		return nil, err
 	}
-	return newCopReqSenderPool(b.ctx, copCtx, sessCtx.GetStore(), b.taskCh, b.checkpointMgr), nil
+	return newCopReqSenderPool(b.ctx, copCtx, sessCtx.GetStore(), b.taskCh, b.sessPool, b.checkpointMgr), nil
 }
 
 func (b *ingestBackfillScheduler) expectedWorkerSize() (readerSize int, writerSize int) {

--- a/ddl/export_test.go
+++ b/ddl/export_test.go
@@ -18,6 +18,7 @@ import (
 	"context"
 	"time"
 
+	"github.com/pingcap/tidb/ddl/internal/session"
 	"github.com/pingcap/tidb/kv"
 	"github.com/pingcap/tidb/sessionctx/stmtctx"
 	"github.com/pingcap/tidb/sessionctx/variable"
@@ -51,7 +52,8 @@ func FetchChunk4Test(copCtx *copContext, tbl table.PhysicalTable, startKey, endK
 	}
 	taskCh := make(chan *reorgBackfillTask, 5)
 	resultCh := make(chan idxRecResult, 5)
-	pool := newCopReqSenderPool(context.Background(), copCtx, store, taskCh, nil)
+	sessPool := session.NewSessionPool(nil, store)
+	pool := newCopReqSenderPool(context.Background(), copCtx, store, taskCh, sessPool, nil)
 	pool.chunkSender = &resultChanForTest{ch: resultCh}
 	pool.adjustSize(1)
 	pool.tasksCh <- task

--- a/ddl/index_cop.go
+++ b/ddl/index_cop.go
@@ -127,7 +127,7 @@ func (c *copReqSender) run() {
 		logutil.BgLogger().Info("[ddl-ingest] start a cop-request task",
 			zap.Int("id", task.id), zap.String("task", task.String()))
 
-		startTS, err := calculateStartTS(se)
+		startTS, err := beginAndGetStartTS(se)
 		if err != nil {
 			p.chunkSender.AddTask(idxRecResult{id: task.id, err: err})
 			return
@@ -513,7 +513,7 @@ type idxRecResult struct {
 	done  bool
 }
 
-func calculateStartTS(se *sess.Session) (uint64, error) {
+func beginAndGetStartTS(se *sess.Session) (uint64, error) {
 	se.Rollback()
 	err := se.Begin()
 	if err != nil {

--- a/ddl/index_cop.go
+++ b/ddl/index_cop.go
@@ -94,9 +94,8 @@ type copReqSender struct {
 func (c *copReqSender) run() {
 	p := c.senderPool
 	defer p.wg.Done()
-	var curTaskID int
 	defer util.Recover(metrics.LabelDDL, "copReqSender.run", func() {
-		p.chunkSender.AddTask(idxRecResult{id: curTaskID, err: dbterror.ErrReorgPanic})
+		p.chunkSender.AddTask(idxRecResult{err: dbterror.ErrReorgPanic})
 	}, false)
 	sessCtx, err := p.sessPool.Get()
 	if err != nil {
@@ -105,10 +104,7 @@ func (c *copReqSender) run() {
 		return
 	}
 	se := sess.NewSession(sessCtx)
-	defer func() {
-		se.Rollback()
-		p.sessPool.Put(sessCtx)
-	}()
+	defer p.sessPool.Put(sessCtx)
 	for {
 		if util.HasCancelled(c.ctx) {
 			return
@@ -123,19 +119,22 @@ func (c *copReqSender) run() {
 				zap.String("task end key", hex.EncodeToString(task.endKey)))
 			continue
 		}
-		curTaskID = task.id
-		logutil.BgLogger().Info("[ddl-ingest] start a cop-request task",
-			zap.Int("id", task.id), zap.String("task", task.String()))
-
-		startTS, err := beginAndGetStartTS(se)
+		err := scanRecords(p, task, se)
 		if err != nil {
 			p.chunkSender.AddTask(idxRecResult{id: task.id, err: err})
 			return
 		}
+	}
+}
+
+func scanRecords(p *copReqSenderPool, task *reorgBackfillTask, se *sess.Session) error {
+	logutil.BgLogger().Info("[ddl-ingest] start a cop-request task",
+		zap.Int("id", task.id), zap.String("task", task.String()))
+
+	return wrapInBeginRollback(se, func(startTS uint64) error {
 		rs, err := p.copCtx.buildTableScan(p.ctx, startTS, task.startKey, task.excludedEndKey())
 		if err != nil {
-			p.chunkSender.AddTask(idxRecResult{id: task.id, err: err})
-			return
+			return err
 		}
 		failpoint.Inject("mockCopSenderPanic", func(val failpoint.Value) {
 			if val.(bool) {
@@ -150,10 +149,9 @@ func (c *copReqSender) run() {
 			srcChk := p.getChunk()
 			done, err = p.copCtx.fetchTableScanResult(p.ctx, rs, srcChk)
 			if err != nil {
-				p.chunkSender.AddTask(idxRecResult{id: task.id, err: err})
 				p.recycleChunk(srcChk)
 				terror.Call(rs.Close)
-				return
+				return err
 			}
 			if p.checkpointMgr != nil {
 				p.checkpointMgr.UpdateTotal(task.id, srcChk.NumRows(), done)
@@ -165,7 +163,22 @@ func (c *copReqSender) run() {
 			p.chunkSender.AddTask(idxRs)
 		}
 		terror.Call(rs.Close)
+		return nil
+	})
+}
+
+func wrapInBeginRollback(se *sess.Session, f func(startTS uint64) error) error {
+	err := se.Begin()
+	if err != nil {
+		return errors.Trace(err)
 	}
+	defer se.Rollback()
+	var startTS uint64
+	sessVars := se.GetSessionVars()
+	sessVars.TxnCtxMu.Lock()
+	startTS = sessVars.TxnCtx.StartTS
+	sessVars.TxnCtxMu.Unlock()
+	return f(startTS)
 }
 
 func newCopReqSenderPool(ctx context.Context, copCtx *copContext, store kv.Storage,
@@ -511,18 +524,4 @@ type idxRecResult struct {
 	chunk *chunk.Chunk
 	err   error
 	done  bool
-}
-
-func beginAndGetStartTS(se *sess.Session) (uint64, error) {
-	se.Rollback()
-	err := se.Begin()
-	if err != nil {
-		return 0, errors.Trace(err)
-	}
-	var startTS uint64
-	sessVars := se.GetSessionVars()
-	sessVars.TxnCtxMu.Lock()
-	startTS = sessVars.TxnCtx.StartTS
-	sessVars.TxnCtxMu.Unlock()
-	return startTS, nil
 }

--- a/ddl/index_cop.go
+++ b/ddl/index_cop.go
@@ -22,6 +22,7 @@ import (
 	"github.com/pingcap/errors"
 	"github.com/pingcap/failpoint"
 	"github.com/pingcap/tidb/ddl/ingest"
+	sess "github.com/pingcap/tidb/ddl/internal/session"
 	"github.com/pingcap/tidb/distsql"
 	"github.com/pingcap/tidb/expression"
 	"github.com/pingcap/tidb/kv"
@@ -70,6 +71,7 @@ type copReqSenderPool struct {
 	tasksCh       chan *reorgBackfillTask
 	chunkSender   chunkSender
 	checkpointMgr *ingest.CheckpointManager
+	sessPool      *sess.Pool
 
 	ctx    context.Context
 	copCtx *copContext
@@ -96,6 +98,17 @@ func (c *copReqSender) run() {
 	defer util.Recover(metrics.LabelDDL, "copReqSender.run", func() {
 		p.chunkSender.AddTask(idxRecResult{id: curTaskID, err: dbterror.ErrReorgPanic})
 	}, false)
+	sessCtx, err := p.sessPool.Get()
+	if err != nil {
+		logutil.BgLogger().Error("[ddl-ingest] copReqSender get session from pool failed", zap.Error(err))
+		p.chunkSender.AddTask(idxRecResult{err: err})
+		return
+	}
+	se := sess.NewSession(sessCtx)
+	defer func() {
+		se.Rollback()
+		p.sessPool.Put(sessCtx)
+	}()
 	for {
 		if util.HasCancelled(c.ctx) {
 			return
@@ -113,12 +126,13 @@ func (c *copReqSender) run() {
 		curTaskID = task.id
 		logutil.BgLogger().Info("[ddl-ingest] start a cop-request task",
 			zap.Int("id", task.id), zap.String("task", task.String()))
-		ver, err := p.store.CurrentVersion(kv.GlobalTxnScope)
+
+		startTS, err := calculateStartTS(se)
 		if err != nil {
 			p.chunkSender.AddTask(idxRecResult{id: task.id, err: err})
 			return
 		}
-		rs, err := p.copCtx.buildTableScan(p.ctx, ver.Ver, task.startKey, task.excludedEndKey())
+		rs, err := p.copCtx.buildTableScan(p.ctx, startTS, task.startKey, task.excludedEndKey())
 		if err != nil {
 			p.chunkSender.AddTask(idxRecResult{id: task.id, err: err})
 			return
@@ -155,7 +169,8 @@ func (c *copReqSender) run() {
 }
 
 func newCopReqSenderPool(ctx context.Context, copCtx *copContext, store kv.Storage,
-	taskCh chan *reorgBackfillTask, checkpointMgr *ingest.CheckpointManager) *copReqSenderPool {
+	taskCh chan *reorgBackfillTask, sessPool *sess.Pool,
+	checkpointMgr *ingest.CheckpointManager) *copReqSenderPool {
 	poolSize := copReadChunkPoolSize()
 	srcChkPool := make(chan *chunk.Chunk, poolSize)
 	for i := 0; i < poolSize; i++ {
@@ -169,6 +184,7 @@ func newCopReqSenderPool(ctx context.Context, copCtx *copContext, store kv.Stora
 		senders:       make([]*copReqSender, 0, variable.GetDDLReorgWorkerCounter()),
 		wg:            sync.WaitGroup{},
 		srcChkPool:    srcChkPool,
+		sessPool:      sessPool,
 		checkpointMgr: checkpointMgr,
 	}
 }
@@ -495,4 +511,18 @@ type idxRecResult struct {
 	chunk *chunk.Chunk
 	err   error
 	done  bool
+}
+
+func calculateStartTS(se *sess.Session) (uint64, error) {
+	se.Rollback()
+	err := se.Begin()
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	var startTS uint64
+	sessVars := se.GetSessionVars()
+	sessVars.TxnCtxMu.Lock()
+	startTS = sessVars.TxnCtx.StartTS
+	sessVars.TxnCtxMu.Unlock()
+	return startTS, nil
 }

